### PR TITLE
PydanticGen CLI Updates

### PR DESCRIPTION
**Pull request by glass-ships**
- Adds `--allow-extra` as a CLI option to make use of Pydantic's "allow extra fields" configuration
- Updates CLI options to use dashes rather than underscores, to match convention and the rest of linkml generators

### CODE REVIEW

**Comment by glass-ships** [comment]
I do wonder if this should instead of being a flag, be a text argument to allow the different options:
```
--extra allow|forbid|ignore
```

**Comment by cmungall** [comment]
What's the difference between allow and ignore?

Consider using consistent options with json schema validation; e.g `--closed`, see #1404.  

Ultimately I think this needs to be configured at the schema level:

- #1595

**Comment by glass-ships** [comment]
> What's the difference between allow and ignore?

allow: extra (unspecified fields) will be included in the instance of a class
ignore: extra fields will not be included, but will also not fail validation

> Consider using consistent options with json schema validation; e.g `--closed`

I mainly named it this way to be consistent with Pydantic's naming of the option/configuration

> Ultimately I think this needs to be configured at the schema level

I considered this too, but I wonder if it's something that makes sense for ..well, pretty much any other generator? 
It totally can be though if that's the more sensible route (though I'm less immediately sure how to implement it that way)

**Comment by cmungall** [comment]
@glass-ships can you get this to pass the unit tests?

**Comment by glass-ships** [comment]
Thanks for the ping @cmungall , i'll take a look at that now! 

One question/suggestion - should we add test output directories to the git ignore? or is that important content to exist in the repo? (it does mean changing one file and running tests locally results in a _lot_ of files modified and showing in a git diff, meaning lots of careful adding of modified files before committing)

**Comment by pkalita-lbl** [comment]
@glass-ships you're absolutely right that the test output directories are cumbersome to work with. Unfortunately they are needed for some of the existing `unittest`-style tests. We have better ways to implement those tests so that they don't modify committed files using `pytest`, but we haven't yet finished the conversion of the entire test suite (see: https://github.com/linkml/linkml/issues?q=is%3Aissue+is%3Aopen+%22convert+tests+in%22).

I tend to stage (`git add`) the changes I intend to commit, run the tests, and then revert any _unstaged_ changes that were caused by running the tests.

**Comment by dalito** [comment]
@pkalita-lbl Is there a reason why you don't write test output to temporary dirs and let pytest manage them?

**Comment by pkalita-lbl** [comment]
That's exactly one of the aspects of converting tests to `pytest` (see for example: https://github.com/linkml/linkml/issues/1715)

**Comment by pkalita-lbl** [comment]
I put a breakpoint at line 41 of `test_compile_pydantic` (in `tests/test_generators/test_pydanticgen.py`) so that I could see the generated code (in the `code` variable) before it goes into the `compile_python` function. When I print out `code` I see it contains:

```python
class ConfiguredBaseModel(WeakRefShimBaseModel,
                validate_assignment = True,
                validate_all = True,
                underscore_attrs_are_private = True,
                extra = forbid,
                arbitrary_types_allowed = True,
                use_enum_values = True):
    pass
```

I assume the issue is that `forbid` needs to be in quotes so it's not interpreted as a variable?

**Comment by glass-ships** [comment]
Ah, it does indeed! `PydanticGenerator` accepts `extra_fields: str`, i guess i wasn't entirely expecting it to come out of `serialize()`... Recommendations on how to approach?

File: linkml/generators/pydanticgen.py [hunk 1]
```diff
@@ -64,26 +64,26 @@ def default_template(pydantic_ver: str = "1") -> str:
...
+        template += f"""
 class WeakRefShimBaseModel(BaseModel):
    __slots__ = '__weakref__'
 
 class ConfiguredBaseModel(WeakRefShimBaseModel,
                 validate_assignment = True,
                 validate_all = True,
                 underscore_attrs_are_private = True,
-                extra = {% if allow_extra %}'allow'{% else %}'forbid'{% endif %},
+                extra = {extra_fields},
```
[the remark this answers is not in the document]
Comment (glass-ships) [reply linkml/generators/pydanticgen.py]: OH!!! oh man, thank you - that fixed most of them. 
a couple leftover errors, but 328 down, 2 to go!

**Comment by glass-ships** [comment]
(Just dropping these notes off so I don't forget between now and after dinner)
Only 2 tests failing left:
```
=========================================================================== short test summary info ===========================================================================
FAILED tests/test_compliance/test_core_compliance.py::test_date_types[python_dataclasses-date-20210101-False] - Failed: DID NOT RAISE <class 'Exception'>
FAILED tests/test_compliance/test_core_compliance.py::test_date_types[jsonschema-date-20210101-False] - AssertionError: Expected errors in json schema validation, but none found
==================================== 2 failed, 3788 passed, 509 skipped, 4 xfailed, 532 warnings, 245 subtests passed in 607.82s (0:10:07) ====================================
```

**Comment by pkalita-lbl** [comment]
If you got those running the tests locally on Python 3.11 that's a separate issue: https://github.com/linkml/linkml/issues/1661. I think if you just [do a reformat](https://linkml.io/linkml/contributing/contributing.html#code-formatting-and-linting) the GitHub testing workflow should pass.

**Comment by glass-ships** [comment]
ah yeah, that was totally it! thanks so much for all your help getting this patched up i appreciate it!

**Review by pkalita-lbl — APPROVED** [review]
I couldn't help but notice an egregious bug that I made in `validation_context.py` -- writing `*closed` instead of `*, closed` in the parameter list. So I hope you don't mind that a snuck in a fix for that and added a more specific test for the `closed` parameter of `PydanticValidationPlugin`. 

I'm fine with the code changes themselves. I'll leave it up to @cmungall if he has any other comments about the semantics regarding LinkML's traditional open/closed vs Pydantic's three-level allow/ignore/forbid.

**Review by cmungall — APPROVED** [review]

**Pull request merged**